R/tostring.R

```diff
@@ -571,77 +598,114 @@ new_line_warning <- function(str_v) {
+#'
+#' @param str character. String to be wrapped. If it is a character vector or
+#'   a list, it will be looped as a list and returned with `unlist(use.names = FALSE)`.
+#' @param width numeric(1). Width, in characters, that the
+#'   text should be wrapped at.
+#' @param collapse character(1) or `NULL`. If the words that have been split should
+#'   be pasted together with the collapse character. This is usually done internally
+#'   with `"\n"` to have the wrapping updated along with other internal values
+#'
+#' @details Word wrapping happens as with [stringi::stri_wrap()]
 #'   with the following exception: individual words which are longer
 #'   than `max_width` are broken up in a way that fits with the rest of the
 #'   word wrapping.
 #'
-#' @return A string (`wrap_string` or character vector (`wrap_txt`) containing
-#'   the hard or soft word-wrapped content.
+#' @return A string if `str` is one element and if `collapse = NULL`. Otherwise, is
+#'   a list of elements (if `length(str) > 1`) that can contain strings or vector of
+#'   characters (if `collapse = NULL`).
+#'
+#' @examples
+#' str <- list("  , something really  \\tnot  very good", # \t needs to be escaped
+#'             "  but I keep it12   ")
+#' wrap_string(str, 5, collapse = "\n")
 #'
+#' @name wrap_string
 #' @export
-wrap_string <- function(str, max_width, hard = FALSE) {
-  stopifnot(is.character(str) && length(str) == 1)
-  naive <- strwrap(str, max_width + 1)
-  while (any(nchar(naive) > max_width)) {
-    good <- character()
-    bwi <- which(nchar(naive) > max_width)[1]
-    curbw <- naive[bwi]
-    if (bwi > 2) {
-      good <- c(good, naive[1:(bwi - 2)])
-    }
-    if (bwi > 1) {
-      str_before <- naive[bwi - 1]
-    } else {
-      str_before <- ""
-    }
-    room <- max_width - nchar(str_before) - (bwi > 1)
-    if (room <= 0) {
-      toadd <- c(str_before, substr(curbw, 1, max_width))
-      room <- 0
-      leftover <- substr(curbw, max_width + 1, nchar(curbw))
-    } else {
-      goodpart <- substr(curbw, 1, room)
-      if (nzchar(str_before)) {
-        toadd <- paste(str_before, goodpart)
-      } else {
-        toadd <- goodpart
-      }
-      leftover <- substr(curbw, room + 1, nchar(curbw))
-    }
-    good <- c(good, toadd)
-    if (bwi == length(naive)) {
-      good <- c(good, leftover)
-    } else {
-      good <- c(
-        good,
-        paste(leftover, naive[bwi + 1]),
-        if (bwi < length(naive) - 1) naive[seq(bwi + 2, length(naive))]
+wrap_string <- function(str, width, collapse = NULL) {
+  if (length(str) > 1) {
+    return(
+      unlist(
+        lapply(str, wrap_string, width = width, collapse = collapse),
+        use.names = FALSE
       )
-    }
-    str <- paste(good, collapse = " ")
-    naive <- strwrap(str, max_width + 1)
+    )
+  }
+  str <- unlist(str, use.names = FALSE) # it happens is one list element
+  if (length(str) == 0 || is.null(str) || !nzchar(str) || is.na(str)) {
+    return(str)
```

Comment:
do you think there is something smarter here? ahah

Reply:
You could simplify a bit to this since `length(NULL) = 0`:
```suggestion
  if (!length(str) || !nzchar(str) || is.na(str)) {
    return(str)
```